pageserver/src/billing_metrics.rs

```diff
@@ -0,0 +1,285 @@
+                tenant_id, timeline.timeline_id, timeline_size, timeline_written_size)
+        }
+
+        let tenant_remote_size = tenant.get_remote_size().await?;
+        debug!(
+            "collected current metrics for tenant: {}: state={:?} tenant_physical_size={} remote_size={}",
+            tenant_id, tenant_state, tenant_physical_size, tenant_remote_size
+        );
+
+        current_metrics.push((
+            BillingMetricsKey {
+                tenant_id,
+                timeline_id: None,
+                metric: BillingMetricKind::PhysicalSize,
+            },
+            tenant_physical_size,
+        ));
+
+        current_metrics.push((
+            BillingMetricsKey {
+                tenant_id,
+                timeline_id: None,
+                metric: BillingMetricKind::S3StorageSize,
+            },
+            tenant_remote_size,
+        ));
+
+        // TODO add SyntheticStorageSize metric
+    }
+
+    // Filter metrics
+    let mut cached_metrics_guard = cached_metrics.lock().await;
+    let mut filtered_metrics = Vec::new();
+
+    for (curr_key, curr_val) in current_metrics.iter() {
+        if let Some(val) = cached_metrics_guard.insert(curr_key.clone(), *curr_val) {
+            if val != *curr_val {
+                // metric was updated, send it
+                filtered_metrics.push({
+                    BillingMetric::new_absolute(
+                        curr_key.metric,
+                        curr_key.tenant_id,
+                        curr_key.timeline_id,
+                        *curr_val,
+                    )
+                });
+            }
+        } else {
+            // cache the metric
+            cached_metrics_guard.insert(curr_key.clone(), *curr_val);
+            filtered_metrics.push({
+                BillingMetric::new_absolute(
+                    curr_key.metric,
+                    curr_key.tenant_id,
+                    curr_key.timeline_id,
+                    *curr_val,
+                )
+            });
+        }
+    }
+
+    if filtered_metrics.is_empty() {
+        trace!("no new metrics to send");
+        return Ok(());
+    }
+
+    // Send metrics to billing service.
+    // split into chunks of 1000 metrics to avoid exceeding the max request size
+    const CHUNK_SIZE: usize = 1000;
+    let chunks = filtered_metrics.chunks(CHUNK_SIZE);
+
+    for chunk in chunks {
+        let chunk_json = serde_json::value::to_raw_value(&EventChunk { events: chunk })
+            .expect("BillingMetric should not fail serialization");
+
+        let res = client
+            .post(metric_collection_endpoint)
+            .json(&chunk_json)
+            .send()
+            .await;
+
+        match res {
+            Ok(res) => {
+                if res.status().is_success() {
+                    debug!("metrics sent successfully, response: {:?}", res);
+                } else {
+                    error!("failed to send metrics: {:?}", res);
```

Comment:
Consider a different kind of string to distinguish these from the "sending failed" cases (dns, network, ...).

```suggestion
                    error!("metrics endpoint refused the send metrics: {:?}", res);
```

Reply:
thank you. Fixed